File: output_20260320_065752/deepseek/c/Struct/deposit.c
#include "stdlib.h"

struct account
{
    int balance;
};

/*@
predicate account(struct account *a;) = a->balance |-> _;
@*/

struct account *create_account()
//@ requires true;
//@ ensures account(result) &*& malloc_block_account(result);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    //@ close account(myAccount);
    myAccount->balance = 0;
    return myAccount;
}

void account_deposit(struct account *myAccount, int amount)
//@ requires account(myAccount);
//@ ensures account(myAccount);
{
    //@ open account(myAccount);
    myAccount->balance += amount;
    //@ close account(myAccount);
}

int account_get_balance(struct account *myAccount)
//@ requires account(myAccount);
//@ ensures account(myAccount);
{
    //@ open account(myAccount);
    int result = myAccount->balance;
    //@ close account(myAccount);
    return result;
}

void account_set_balance(struct account *myAccount, int newBalance)
//@ requires account(myAccount);
//@ ensures account(myAccount);
{
    //@ open account(myAccount);
    myAccount->balance = newBalance;
    //@ close account(myAccount);
}

void account_dispose(struct account *myAccount)
//@ requires account(myAccount) &*& malloc_block_account(myAccount);
//@ ensures true;
{
    //@ open account(myAccount);
    free(myAccount);
}

int main()
//@ requires true;
//@ ensures true;
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    //@ assert b == 15;
    account_dispose(myAccount);
    return 0;
}

File: output_20260320_065752/deepseek/rust/Struct/deposit.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Account {
    balance: i32,
}

/*@
pred_ctor Account_own(account: *mut Account, balance: i32)() = 
    struct_Account { balance }@account;
@*/

impl Account {
    //@ req true;
    //@ ens Account_own(result, 0)();
    unsafe fn create() -> *mut Account
    
    
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        //@ close Account_own(my_account, 0)();
        (*my_account).balance = 0;
        return my_account;
    }

    //@ req Account_own(my_account, ?balance)();
    //@ ens Account_own(my_account, balance)() &*& result == balance;
    unsafe fn get_balance(my_account: *mut Account) -> i32
    
    
    {
        //@ open Account_own(my_account, ?balance)();
        let result = (*my_account).balance;
        //@ close Account_own(my_account, balance)();
        return result;
    }

    //@ req Account_own(my_account, _)();
    //@ ens Account_own(my_account, new_balance)();
    unsafe fn set_balance(my_account: *mut Account, new_balance: i32)
    
    
    {
        //@ open Account_own(my_account, _)();
        (*my_account).balance = new_balance;
        //@ close Account_own(my_account, new_balance)();
    }

    //@ req Account_own(my_account, ?balance)();
    //@ ens Account_own(my_account, balance + amount)();
    unsafe fn deposit(my_account: *mut Account, amount: i32)
    
    
    {
        //@ open Account_own(my_account, ?balance)();
        (*my_account).balance += amount;
        //@ close Account_own(my_account, balance + amount)();
    }

    //@ req Account_own(my_account, _)();
    //@ ens true;
    unsafe fn dispose(my_account: *mut Account)
    
    
    {
        //@ open Account_own(my_account, _)();
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }

}

fn main()

{
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}